cmd/gitlab.go

```diff
@@ -9,303 +9,77 @@ import (
-		log.Debugf("Found %d refs from %s pipelines", len(refs), p.Name)
-		for _, r := range refs {
-			log.Infof("Found ref '%s' for project '%s'", r, p.Name)
-			if err := c.pollProjectRef(p, r); err != nil {
-				log.Errorf("Error getting pipeline data on ref '%s' for project '%s'", r, p.Name)
-				continue
-			}
-			polledRefs = append(polledRefs, r)
-		}
-	}
-	log.Debugf("Polling refs for project : %s", p.Name)
-
-	refs, err := c.branchesAndTagsFor(p.GitlabProject.ID, p.Refs)
+	branchesAndTagRefs, err := c.branchesAndTagsFor(p.GitlabProject.ID, p.Refs)
 	if err != nil {
 		return fmt.Errorf("error fetching refs for project '%s'", p.Name)
 	}
-
-	if len(refs) > 0 {
-		for _, r := range refs {
-			if !refExists(polledRefs, r) {
-				log.Infof("Found ref '%s' for project '%s'", r, p.Name)
-				if err := c.pollProjectRef(p, r); err != nil {
-					log.Errorf("Error getting pipeline data on ref '%s' for project '%s'", r, p.Name)
-					continue
-				}
-				polledRefs = append(polledRefs, r)
-			}
-		}
+	if len(branchesAndTagRefs) == 0 {
+		log.Warnf("No refs found for project '%s'", p.Name)
+		return nil
 	}
-
-	log.Warnf("No refs found for project '%s'", p.Name)
-	return nil
-}
-
-func outputStatusMetric(metric *prometheus.GaugeVec, labels []string, statuses []string, status string, sparseMetrics bool) {
-	// Moved into separate function to reduce cyclomatic complexity
-	// List of available statuses from the API spec
-	// ref: https://docs.gitlab.com/ee/api/jobs.html#list-pipeline-jobs
-	for _, s := range statuses {
-		args := append(labels, s)
-		if s == status {
-			metric.WithLabelValues(args...).Set(1)
-		} else {
-			if sparseMetrics {
-				metric.DeleteLabelValues(args...)
-			} else {
-				metric.WithLabelValues(args...).Set(0)
-			}
+	// read the metrics for refs
+	log.Debugf("Polling refs for project : %s", p.Name)
+	for _, r := range branchesAndTagRefs {
+		log.Infof("Found ref '%s' for project '%s'", r, p.Name)
+		if err := c.pollProjectRefOn(project, r, p.ShouldOutputSparseStatusMetrics(cfg), p.ShouldFetchPipelineJobMetrics(cfg)); err != nil {
+			log.Errorf("Error getting pipeline data on ref '%s' for project '%s'", r, p.Name)
+			continue
 		}
 	}
-}
 
-func (c *Client) outputPipelineStatusMetric(status string, sparseMetrics bool, labels ...string) {
-	outputStatusMetric(
-		lastRunStatus,
-		labels,
-		[]string{"running", "pending", "success", "failed", "canceled", "skipped", "manual"},
-		status,
-		sparseMetrics,
-	)
-}
-func (c *Client) outputJobStatusMetric(status string, sparseMetrics bool, labels ...string) {
-	outputStatusMetric(
-		lastRunJobStatus,
-		labels,
-		[]string{"running", "pending", "success", "failed", "canceled", "skipped", "manual"},
-		status,
-		sparseMetrics,
-	)
+	return nil
 }
 
-func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref string) error {
+func (c *Client) pollPipelineJobs(gp *gitlab.Project, pipelineID int, topics string, ref string, outputSparseMetricsStatus bool) error {
```

Comment:
I see that you are using `gitlabProject` in another function: https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/pull/93/files#diff-fb2ba0b24813a3ba03c719c857bd28d1R141

I do not mind the terminology but I reckon it would be nice to have a single way to name this object across the codebase

Reply:
Agreed! renamed to `gp` to keep it consistent 👍